rfc/20240513-static-evaluation-providers.md

```diff
@@ -288,34 +442,39 @@ Example (with relevant fields only):
+  "type": "aws_instance",
+  "name": "example",
+  "provider": "module.example.provider[\"registry.opentofu.org/hashicorp/aws\"].by_region",
   "instances": [
     {
       "index_key": "first",
-      "provider": "module.dummy.provider[\"registry.opentofu.org/hashicorp/tfcoremock\"].firstalias"
+      "provider": "module.example.provider[\"registry.opentofu.org/hashicorp/aws\"].by_region[\"us-west-2\"]"
     },
     {
       "index_key": "second",
-      "provider": "module.dummy.provider[\"registry.opentofu.org/hashicorp/tfcoremock\"].secondalias"
+      "provider": "module.example.provider[\"registry.opentofu.org/hashicorp/aws\"].by_region[\"eu-east-1\"]"
     }
   ]
 }
 ```
 
-Inspecting the history of the state file versioning, this fits well within changes allowed in the same version. This would break compatibility for downgrading to older OpenTofu versions and should be noted in the changelog. It is possible to migrate back to not using provider iteration, but that will need to be applied before any downgrade could occur.
+The state snapshot loader will parse both the resource-level address and the instance-level addresses and will, for now, return an error if they are not identical aside from the optional trailing instance key on the instance-level addresses. This preserves our current simplification of each resource still being bound to exactly one provider configuration while offering a more general syntax that we can retain in future if we weaken that constraint. We don't yet know exactly what flexibility we will need for future features, and so capturing the entire provider instance address (and then verifying their consistency) gives us the freedom to refer to literally any valid provider instance address in future versions, without changing the syntax and thus without the need for new state snapshot format upgrade logic.
+
+Retaining the `"provider"` argument at the resource level, despite it now being technically redundant, offers a measure of backward-compatibility with older versions of OpenTofu. Someone who upgrades to the version that introduced this feature and runs `tofu apply` should remain able to revert back to the previous OpenTofu version as long as their configuration does not make use of provider configuration expansion, because the older version will ignore the instance-level `"provider"` properties altogether.
 
-Note: This only applies to the state file, `tofu state show -json` has it's own format which is not impacted by this change.
+This section described changes only to the internal state snapshot format that is not documented for external use. We do not intend to extend the documented `tofu show -json` output formats yet as part of this change, because we want to retain design flexibility for later work and to learn more about what real-world use-cases exist for machine-consumption of the provider instance relationships before we design a final format that will then be subject to compatibilty constraints.
```

Comment:
That field has a slightly different meaning now.

Previously, third parties could match resources with providers via a single field, but now resource instance providers should also be taken into account. This could cause unexpected behavior for those tools, and I would prefer them to fail instead of producing irrelevant results.

I know third parties should not read a state file directly, however some tools are still doing it.

Also, it reduces the risk of unexpected behavior when you run an older tofu version with an updated state file. It might not be the case if tofu fails when encounters unknown state fields.

What do you think about removing this field for cases, when instance level provider should be used?

Reply:
In a synchronous discussion today I believe we resolved that:

- The resource-level `"provider"` property and the instance-level `"provider"` properties will be treated as mutually-exclusive: any resource that has instance-level `"provider"` will _not_ have the resource-level `"provider"`, with the intention of causing a louder failure in any existing external tool that tries to consume the state snapshot format despite it not being a committed public interface.
- OpenTofu will include the instance-level `"provider"` properties only if the instances of a particular resource are associated with instances of a multi-instance `provider` block. Otherwise it will continue to use the resource-level `"provider" property exclusively, to allow rolling back to older OpenTofu versions as long as dynamic provider instance selection hasn't been used yet.
- If instance-level `"provider"` properties are present then they must still currently all agree on everything except the provider instance key. State snapshot loading will fail if the addresses are inconsistent, because the language runtime will continue to rely on each `resource` block being associated with exactly one `provider` block for now.

I think there's still one finicky question to answer before we commit to that as the right answer: if someone uses the `-target=` option to select only one of multiple instances of a resource, could the fact that we're only visiting a subset of the resource instances cause the resulting state to violate the consistency rules described above? If so, how should we resolve that?